contract/models/contract.py

```diff
@@ -0,0 +1,427 @@
+        if self.contract_type == 'purchase':
+            self.payment_term_id = \
+                self.partner_id.property_supplier_payment_term_id
+        else:
+            self.payment_term_id = \
+                self.partner_id.property_payment_term_id
+        self.invoice_partner_id = self.partner_id.address_get(['invoice'])[
+            'invoice'
+        ]
+        return {
+            'domain': {
+                'invoice_partner_id': [
+                    '|',
+                    ('id', 'parent_of', self.partner_id.id),
+                    ('id', 'child_of', self.partner_id.id),
+                ]
+            }
+        }
+
+    def _convert_contract_lines(self, contract):
+        self.ensure_one()
+        new_lines = self.env['contract.line']
+        contract_line_model = self.env['contract.line']
+        for contract_line in contract.contract_line_ids:
+            vals = contract_line._convert_to_write(contract_line.read()[0])
+            # Remove template link field
+            vals.pop('contract_template_id', False)
+            vals['date_start'] = fields.Date.context_today(contract_line)
+            vals['recurring_next_date'] = fields.Date.context_today(
+                contract_line
+            )
+            new_lines += contract_line_model.new(vals)
+        new_lines._onchange_date_start()
+        new_lines._onchange_is_auto_renew()
+        return new_lines
+
+
+    def _prepare_invoice(self, date_invoice, journal=None):
+        """
+        Taken from sale/models/sale.py
+        Prepare the dict of values to create the new invoice for a sales order. This method may be
+        overridden to implement custom invoice generation (making sure to call super() to establish
+        a clean extension chain).
+        """
+        self.ensure_one()
+        if self.contract_type == 'sale':
+            invoice_type = 'out_invoice'
+        else: #purchase
+            invoice_type = 'in_invoice'
+        if not journal:
+            if self.journal_id:
+                journal = self.journal_id
+            else:
+                journal = self.env['account.move'].with_context(force_company=self.company_id.id, default_type=invoice_type)._get_default_journal()
+                if not journal:
+                    raise UserError(_('Please define an accounting %s journal for the company %s (%s).') % (self.contract_type, self.company_id.name, self.company_id.id))
+
+        currency = (
+            self.pricelist_id.currency_id
+            or self.partner_id.property_product_pricelist.currency_id
+            or self.company_id.currency_id
+        )
+
+        invoice_vals = {
+            'name' : '/',
+            'date': date_invoice,
+            'ref':  '%s contract id=%s'%(self.contract_type,self.id),
+            'type': invoice_type,
+            'journal_id': journal.id,
+            'invoice_origin': 'Contract'+self.name+'id='+str(self.id),
```

Comment:
Should be translatable ?
and thingk nternal id is not a info to use ? self.name should be better ?